### PR TITLE
refactor: `Closure` should use expressions

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@theannanas @ChrBondi @plun411 This PR allows Closures to contain expressions. But it reveals a bug somewhere in the Optimizer which you have to debug.

**Comment by magnus-madsen** [comment]
But this means you should be able to proceed unblocked. 

Do you want me to merge this PR, disabling the optimizer in the process? Or should I wait for you to fix the issue? 

How do you want to proceed?

**Comment by ChrBondi** [comment]
@magnus-madsen i think you should just merge it in.

**Comment by ChrBondi** [comment]
> > @magnus-madsen i think you should just merge it in.
> 
> This will then disable the Optimizer/leave it broken. But then you will fix it?

Yes we will look into it

File: main/src/ca/uwaterloo/flix/language/phase/jvm/ClosureInfo.scala [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
 /**
   * Meta information about a closure.
   */
-case class ClosureInfo(sym: Symbol.DefnSym, freeVars: List[FreeVar], tpe: MonoType) {
+case class ClosureInfo(sym: Symbol.DefnSym, closureArgs: List[Expression], tpe: MonoType) {
   /**
     * Returns the hash code of `this` closure info.
     */
-  override def hashCode(): Int = 7 * sym.hashCode + 11 * freeVars.hashCode()
+  override def hashCode(): Int = 7 * sym.hashCode + 11 * closureArgs.hashCode()
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/ClosureInfo.scala:13]: @JonathanStarup 

I am concerned about this use of hash code on expressions. Why is equality needed on ClosureInfo? How is it used?

Computing hash code of expressions is not really well-defined...

Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/jvm/ClosureInfo.scala]: Only the types of the arguments were needed. the tpe was left out before so I'm not sure about how to compute equality. Would there ever be two different closures with the same symbol?

**Comment by magnus-madsen** [comment]
@theannanas @ChrBondi @plun411 I have merged this. I suggest you quickly debug and create a PR to re-enable the Optimizer :)

**Comment by magnus-madsen** [comment]
@JonathanStarup was kind enough to leave you to places to get started:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Inliner.scala#L151
https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Inliner.scala#L185

**Pull request merged**